cordyceps/src/sorted_list.rs

```diff
@@ -0,0 +1,459 @@
+    /// front, then you could provide a function something like:
+    ///
+    /// ```rust
+    /// let f: fn(&i32, &i32) -> core::cmp::Ordering = |lhs, rhs| {
+    ///     lhs.cmp(rhs)
+    /// };
+    /// ```
+    ///
+    /// SortedList takes a function (and not just [Ord]) so you can use it on types
+    /// that don't have a general way of ordering them, allowing you to select a
+    /// specific metric within the sorting function.
+    ///
+    /// If two items are considered of equal value, new values will be placed AFTER
+    /// old values.
+    pub const fn new_custom(f: fn(&T, &T) -> core::cmp::Ordering) -> Self {
+        Self {
+            func: f,
+            head: None,
+        }
+    }
+
+    /// Create a new sorted list, consuming the stack, using the provided ordering function
+    pub fn from_stack_custom(stack: Stack<T>, f: fn(&T, &T) -> core::cmp::Ordering) -> Self {
+        let mut slist = Self::new_custom(f);
+        slist.extend(stack);
+        slist
+    }
+
+    /// Pop the front-most item from the list, returning it by ownership (if it exists)
+    ///
+    /// This is an `O(1)` operation.
+    #[must_use]
+    pub fn pop_front(&mut self) -> Option<T::Handle> {
+        test_trace!(?self.head, "SortedList::pop_front");
+        let head = self.head.take()?;
+        unsafe {
+            // Safety: we have exclusive ownership over this chunk of stack.
+
+            // advance the head link to the next node after the current one (if
+            // there is one).
+            self.head = T::links(head).as_mut().next.with_mut(|next| (*next).take());
+
+            test_trace!(?self.head, "SortedList::pop -> popped");
+
+            // return the current node
+            Some(T::from_ptr(head))
+        }
+    }
+
+    /// Insert a single item into the list, in it's sorted order position
+    ///
+    /// This is an `O(n)` operation.
+    pub fn insert(&mut self, element: T::Handle) {
+        let ptr = T::into_ptr(element);
+        test_trace!(?ptr, ?self.head, "SortedList::insert");
+
+        // Take a long-lived reference to the new element
+        let eref = unsafe { ptr.as_ref() };
+
+        // Special case for empty head
+        //
+        // If the head is null, then just place the item
+        let Some(mut cursor) = self.head else {
+            // todo: assert element.next is null?
+            self.head = Some(ptr);
+            return;
+        };
+
+        // Special case for head: do we replace current head with new element?
+        {
+            // compare, but make sure we drop the live reference to the cursor
+            let cmp = {
+                let cref = unsafe { cursor.as_ref() };
+                (self.func)(cref, eref)
+            };
+
+            // If cursor node is LESS or EQUAL: keep moving.
+            // If cursor node is GREATER: we need to place the new item BEFORE
+            if cmp == core::cmp::Ordering::Greater {
```

Comment:
nit, take it or leave it: maybe we should just import `core::cmp::Ordering` in this file...

Reply:
Nah we totally can do that!